### PR TITLE
Circular dependencies: Clean up services

### PR DESCRIPTION
**Pull request by garbear**
This breaks two circular dependencies:

* Settings and PowerManager
* Input and Peripherals

It also cleans up a bunch of dependencies on CServiceBroker, itself a cause of circular dependency.

## Motivation and Context
Initial step in breaking circular dependencies for the login screen. That project is too big to tackle, so I'm chipping away at the edges by reducing some circular dependencies among services.

## How Has This Been Tested?
Untested - Kodi compiles and starts on OSX.

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

### CODE REVIEW

**Comment by garbear** [comment]
hm, file upload failed across the board. jenkins build this please

File: xbmc/Application.cpp [hunk 1]
```diff
@@ -2784,7 +2784,7 @@ void CApplication::Stop(int exitCode)
 
     CVariant vExitCode(CVariant::VariantTypeObject);
     vExitCode["exitcode"] = exitCode;
-    CAnnouncementManager::GetInstance().Announce(System, "xbmc", "OnQuit", vExitCode);
+    m_ServiceManager->GetAnnouncementManager().Announce(ANNOUNCEMENT::System, "xbmc", "OnQuit", vExitCode);
```
[the remark this answers is not in the document]
Comment (garbear) [reply xbmc/Application.cpp]: Sorry, forgot. Will fix

File: xbmc/ServiceManager.cpp [hunk 2]
```diff
@@ -176,11 +180,15 @@ bool CServiceManager::InitStageTwo(const CAppParamParser &params)
   m_fileExtensionProvider.reset(new CFileExtensionProvider(*m_addonMgr,
                                                            *m_binaryAddonManager));
 
-  m_powerManager.reset(new CPowerManager());
+  m_powerManager.reset(new CPowerManager(*m_announcementManager,
```
Comment (FernetMenta) [inline xbmc/ServiceManager.cpp]: this hard dependencies are really bad and make things worse. every service should be stoppable without the need to know what other component may depend on it.

File: xbmc/powermanagement/PowerManager.cpp [hunk 3]
```diff
@@ -34,9 +35,17 @@
...
-
-CPowerManager::CPowerManager() = default;
+CPowerManager::CPowerManager(ANNOUNCEMENT::CAnnouncementManager &announcements,
+                             CSettings &settings,
+                             CNetworkBase &network,
+                             PVR::CPVRManager &pvr) :
+  m_announcements(announcements),
+  m_settings(settings),
+  m_network(network),
+  m_pvr(pvr)
```
Comment (FernetMenta) [inline xbmc/powermanagement/PowerManager.cpp]: PVR has ever been an optional component. Creating this hard dependency is not a good idea.

**Comment by FernetMenta** [comment]
-1
this brings us back to where we came from. by using this kind of dependency injection the order of construction of service is important. exactly this is a big problem. 
we discussed this already in the context of GUI. ServiceBroker should return a pointer so that service can come and go at any time.

**Comment by garbear** [comment]
I'm trying to break up our login screen's circular dependencies. No hard dependencies were introduced that weren't already there. I understand your point that making these hard dependencies explicit is the wrong approach, but I need the login screen problem to become more tractable, and making it explicit helps me understand the dependency graph.

Do you have a better idea of how I can approach fixing the login screen?

**Comment by FernetMenta** [comment]
> Do you have a better idea of how I can approach fixing the login screen?

Regarding PowerManager and PVR: There is an announcement for OnSleep. Why does PowerManager call a method of PVR (OnSleep) directly?

PowerManager and network: 
m_network.WaitForNet seems wrong. network can come and go at any time, just pull the cable. Waiting for network is wrong, you might wait forever. Apart from this, network is optional like PVR.

**Comment by ksooo** [comment]
> Regarding PowerManager and PVR: There is an announcement for OnSleep. Why does PowerManager call a method of PVR (OnSleep) directly?

 IIRC, it was you who wanted to have it this way. Can't remember the exact reason, it was something with async communication does not work here.

**Comment by ksooo** [comment]
@FernetMenta discussion sync. vs async.: https://github.com/xbmc/xbmc/pull/11887#issuecomment-288143994

**Comment by FernetMenta** [comment]
Right. Doing announce for OnSleep is a bad idea in general because the sender won't get any feedback and as a result it won't know if clients have finished their tasks.
Client should register like they do for DisplayLost in Windowing.

**Comment by garbear** [comment]
I split out some commits to other PRs

**Comment by garbear** [comment]
Final remaining useful stuff split out to https://github.com/xbmc/xbmc/pull/14336

**Pull request closed without merging**